File: WolfDen/wwwroot/ts/TurnResolver.ts
namespace TurnBasedBoardGame {
    export class TurnResolver {
        Game: Game;
        ActiveGroupIndex: number;
        TurnOrder: AdvanceWarsPlayer[][];
        ActivePlayers: AdvanceWarsPlayer[];
        Id: string;
        StartTurnEvent: any = {};
        StartTurnModal: StartTurnModal;
        IncomeCoordinator: IncomeCoordinator;
        private PlayerEndTurnEvent: ((player: AdvanceWarsPlayer) => void)[] = [];

        SubscribeToPlayerEndTurnEvent(funct: (player: AdvanceWarsPlayer) => void) {
            this.PlayerEndTurnEvent.push(funct);
        }

        UnsubscribeToPlayerEndTurnEvent(funct: (player: AdvanceWarsPlayer) => void) {
            this.PlayerEndTurnEvent.splice(this.PlayerEndTurnEvent.indexOf(funct),1);
        }

        RaisePlayerEndTurnEvent(player: AdvanceWarsPlayer) {
            for (let f of this.PlayerEndTurnEvent)
                f(player);
        }

        constructor() {
            this.StartTurnModal = new StartTurnModal(this);
        }

        public IsActive(player: AdvanceWarsPlayer) {
            return this.ActivePlayers.indexOf(player) != -1;
        }

        EndTurn(player: AdvanceWarsPlayer) {
            this.RaisePlayerEndTurnEvent(player);
            this.ActivePlayers.splice(this.ActivePlayers.indexOf(player), 1);
            if (player.Id === this.Game.CurrentPlayer.Id)
                $(".endTurnButton").first().prop("disabled", true);
            for (let piece of player.Pieces) {
                if ("Active" in piece) {
                    var activatable = piece as IActive;
                    activatable.Active = false;
                }
            }

            if (this.ActivePlayers.length === 0) {

                while (true) {
                    this.ActiveGroupIndex = (this.ActiveGroupIndex + 1) % this.TurnOrder.length;
                    if (this.TurnOrder[this.ActiveGroupIndex].map(p => p.Defeated).indexOf(false) != -1)
                        break;
                }

                for (let player of this.TurnOrder[this.ActiveGroupIndex]) {
                    this.ActivePlayers.push(player);
                }
                for (let player of this.Game.Players)
                    for (let piece of player.Pieces)
                        if ("Active" in piece) {
                            var activatable = piece as IActive;
                            activatable.Active = true;
                        }
                if (this.ActivePlayers.indexOf(this.Game.CurrentPlayer) != -1)
                    $(".endTurnButton").first().prop("disabled", false);
                this.RaiseStartTurnEvent();
                this.StartTurnModal.StartTurn(this.ActivePlayers);                
            }
        }

        RaiseStartTurnEvent() {
            for (let key in this.StartTurnEvent) {
                if (typeof this.StartTurnEvent[key] === "function")
                    this.StartTurnEvent[key](this.ActivePlayers);
            }
        }
    }
}

File: WolfDen/Game/TurnResolver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using WolfDen.Game.EventArguments;
using WolfDen.Game.Interfaces;
using WolfDen.Game.Other;

namespace WolfDen.Game
{
    public class TurnResolver
    {
        public Game Game { get; set; }
        public string Class { get { return GetType().Name; } }
        public int ActiveGroupIndex { get; set; }
        public AdvanceWarsPlayer[][] TurnOrder { get; set; }
        public List<AdvanceWarsPlayer> ActivePlayers { get; set; }
        public event ListOfPlayersEventHandler StartTurnEvent;
        public IncomeCoordinator IncomeCoordinator { get; set; }

        public TurnResolver(List<AdvanceWarsPlayer> players, Game game)
        {
            Game = game;
            var playersGroupedByTeam = players.GroupBy(player => player.Team, player => player);
            TurnOrder = new AdvanceWarsPlayer[playersGroupedByTeam.Count()][];
            var i = 0;
            foreach (var playerGroup in playersGroupedByTeam)
            {
                TurnOrder[i] = playerGroup.ToArray();
                i++;
            }
            IncomeCoordinator = new IncomeCoordinator(this);
            foreach (var player in players)
                player.RepairCoordinator.Initialize(IncomeCoordinator);
        }

        public bool EndTurn(AdvanceWarsPlayer player)
        {
            if (!ActivePlayers.Contains(player))
                return false;
            ActivePlayers.Remove(player);
            var activePieces = player.Pieces.OfType<IActive>();
            foreach (var activePiece in activePieces)
                activePiece.Active = false;
            if (ActivePlayers.Count == 0)
            {
                while (true)
                {
                    ActiveGroupIndex = (ActiveGroupIndex + 1) % TurnOrder.Length;
                    if (TurnOrder[ActiveGroupIndex].Where(p => !p.Defeated).Count() > 0)
                        break;
                }
                StartTurn();
            }
            return true;
        }

        public void StartTurn()
        {
            Game.History.Turn++;
            ActivePlayers = TurnOrder[ActiveGroupIndex].ToList();
            foreach (var player in Game.Players)
                foreach (var piece in player.Pieces)
                {
                    var activatablePiece = piece as IActive;
                    if (activatablePiece != null)
                        activatablePiece.Active = true;
                }
            RaiseStartTurnEvent();
        }

        public virtual void RaiseStartTurnEvent()
        {
            if (StartTurnEvent != null)
                StartTurnEvent(this, new ListOfPlayersEventArgs(TurnOrder[ActiveGroupIndex].ToList()));
        }

        public bool PlayerIsActive(AdvanceWarsPlayer player)
        {
            return ActivePlayers.Contains(player);
        }
    }
}